koap/src/main/kotlin/Decoder.kt

```diff
@@ -84,15 +98,125 @@ inline fun <reified T : Message> ByteArray.decode(): T =
  * ```
  */
 fun ByteArray.decodeUdp(): Message.Udp {
-    val buffer = Buffer().write(this)
+    val header = decodeUdpHeader()
+    return decode(header, offset = header.size)
+}
 
-    // Header
+/**
+ * Decodes [ByteArray] receiver to a TCP [Message].
+ *
+ * [Figure 4: CoAP Frame for Reliable Transports](https://tools.ietf.org/html/rfc8323#section-3.2)
+ * used for TCP [Message]:
+ *
+ * ```
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |  Len  |  TKL  | Extended Length (if any, as chosen by Len) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |      Code     | Token (if any, TKL bytes) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |  Options (if any) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |1 1 1 1 1 1 1 1|    Payload (if any) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * ```
+ */
+fun ByteArray.decodeTcp(): Message.Tcp {
+    val header = decodeTcpHeader()
+    return decode(header, offset = header.size)
+}
+
+/**
+ * Assembles a [Message.Udp] by decoding CoAP message content (Options + Payload) from the
+ * [ByteArray] receiver and combining with provided [header].
+ *
+ * If [ByteArray] receiver contains the CoAP message header, then specify an offset to begin
+ * decoding after header, for example:
+ *
+ * ```
+ * val header = encoded.decodeUdpHeader()
+ * val message = encoded.decode(header, offset = header.size)
+ * ```
+ */
+fun ByteArray.decode(
+    header: Header.Udp,
+    offset: Int = 0
```

Comment:
Would it make more sense to remove the default? Or possibly have the default as `offset: Int = header.size`?

Reply:
I was really torn on this one.
Couldn't decide what would be more commonly used.

On one hand, if you've decoded the header, it's probably that you've discarded the header data and are now only holding onto the "content" (Options + Payload) of the CoAP encoded message. In this case, `offset` of `0` makes the most sense.

I think from an API perspective though, a default of `header.size` makes the most sense (is the most intuitive) when glancing at this function signature.

I'll make the change, thanks for the suggestion!